### PR TITLE
Add support for language selection in preferences

### LINKED ISSUE #1001: How to change UI language
_From [mentat...@gmail.com](https://code.google.com/u/116256814944482205201/) on 2012-04-21T12:22:36Z_

Hi,

I'd like to change the UI language: default from OS is French in my computer, but I want it in English. I thought it was obvious, but I can't find the settings.

Any idea?
Thanks

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=1001_

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #1001 

---

## Description

Add support for language selection in preferences. 

This was a bit trickier than expected, namely the configuration of the language needs to be stored in a separate file, because we cannot use CONF at the stage the `_` function is called.

I added a `langconfig` in the .`spyder` folder, which stores the language value ('en', 'fr' etc)

With this change I had to implement some new logic in the config dialog, so now comboboxes (so far) have an extra parameter call `restart`  (by default False).

## Preferences dialog

image

## Pop up message box
When the user changes one or more options flagged with `restart=True` and then clicks on Apply or Ok, the a message box will popup informing the user which options need to restart and asking if he/she wants to restart right away.

image

If the user accepts, any restart specific (defined) settings will be applied and spyder will restart automatically.

### Note:
The message box is now available for any other combobox that is created with the `restart=True` parameter. Adding this functionality to other widgets will be left for another PR.

## Todo
- [x] Display language names In original Language
- [x] Agree on additional file method and name
- [x] Define when to pop the box to the user for restart

## For another PR
- Add the restart functionality to all the other widgets besides comboboxes

### CODE REVIEW

File: spyderlib/baseconfig.py [hunk 1]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
 #==============================================================================
 # Translations
 #==============================================================================
+LANG_FILE = osp.join(get_conf_path(), '.langconfig')
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: I'm ok with adding this new file, but don't call it `.langconfig`, just `langconfig`, so that it's not hidden on Linux :-)

At the end, it's going to be inside of `~/.spyder`, so there is no need to be hidden.

File: spyderlib/baseconfig.py [hunk 2]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
 # Translations
 #==============================================================================
+LANG_FILE = osp.join(get_conf_path(), '.langconfig')
+DEFAULT_LANGUAGE = 'en'
+
+# This needs to be updated every time a new language is added to spyder.
+LANGUAGE_CODES = {'en': u'English',
+                  'fr': u'Français',
+                  'es': u'Español',
+                  'pt_BR': u'Português (brasileiro)'
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Is Brazilian Portuguese too different from Portuguese? If not, I'd vote to just leave Portuguese here

Comment (goanpeca) [inline spyderlib/baseconfig.py]: We should ask the guy that did the translation. Can you add him to the thread?

File: spyderlib/baseconfig.py (hunk 1 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: :+1:

File: spyderlib/baseconfig.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: But i do prefer to put just portuguese...

File: spyderlib/plugins/configdialog.py [hunk 3]
```diff
@@ -192,6 +199,7 @@ def current_page_changed(self, index):
         self.apply_btn.setEnabled(widget.is_modified)
         
     def add_page(self, widget):
+        setattr(widget, 'parent_config_dialog', self)
```
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: Oops i do no longer need this. Will remove it.

File: spyderlib/baseconfig.py [hunk 4]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
+                  }
+
+
+def get_available_translations():
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: I don't think we need a function for this ;-)

There are so few translations (and none on the works), that we can just create the list of `langs` by hand here.

File: spyderlib/baseconfig.py [hunk 5]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
+    # is added, to ensure LANGUAGE_CODES is updated.
+    for lang in langs:
+        if lang not in LANGUAGE_CODES:
+            error = _('Update LANGUAGE_CODES if a new translation has been '
+                      'added to Spyder')
+            raise Exception(error)
+    return langs
+
+
+def get_interface_language():
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: It also seems like too much to have this function

File: spyderlib/utils/iofuncs.py [hunk 6]
```diff
@@ -362,7 +362,8 @@ def load_dictionary(filename):
                       'spyder.ini', 'temp.py', 'temp.spydata', 'template.py',
                       'history.py', 'history_internal.py', 'workingdir',
                       '.projects', '.spyderproject', '.ropeproject',
-                      'monitor.log', 'monitor_debug.log', 'rope.log')
+                      'monitor.log', 'monitor_debug.log', 'rope.log',
+                      '.langconfig')
```
Comment (ccordoba12) [inline spyderlib/utils/iofuncs.py]: `.langconfig` -> `langconfig`

Comment (goanpeca) [inline spyderlib/utils/iofuncs.py]: Sure

File: spyderlib/baseconfig.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: Yes it might be too much, but why doing stuff by hand, if a function can handle it?

I would really like to have many more translations

File: spyderlib/baseconfig.py (hunk 4 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: Probably we do not need it now, but it is a harmless function, and it provides a check when adding a new language to make sure things are done properly. I think we should leave it

Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Ok, you're right :-)

If you want more translations, you should look at https://www.transifex.com/

I don't know exactly how it works, but I think we need to upload our pot files and volunteers help to translate them to their native languages :-)

File: spyderlib/baseconfig.py (hunk 5 above)
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Ok, now that it's written, we should leave it  :-)

File: spyderlib/baseconfig.py (hunk 4 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: I will take a look a transifex, I had seen it before, but never got the time to go in depth

**Comment by goanpeca** [comment]
@ccordoba12 I updated with the fixes, when you have some time can you take a look at some questions I left in the description?

**Comment by goanpeca** [comment]
Can @spyder-ide/core-developers give this a try and give some comments?

```bash
git clone https://github.com/goanpeca/spyder.git spyder-lang
cd spyder-lang
git checkout -b language
git pull origin language
python bootstrap.py
```

**Comment by goanpeca** [comment]
This PR is waiting for some input from the @spyder-ide/core-developers ...

There are 2 ways of doing this.
1.) Ask the user if he wants to restart for the setting to take place, immediately after changing the setting
2.) Ask the user if he wants to restart for the setting to take place, only after pressing Ok or Apply

I implemented the first one.. but the second one could be also implemented if you consider is better.

@ccordoba12 any ideas, thoughts?

**Comment by Nodd** [comment]
I vote 2 because:
* there could be other settings which ask to reload (changing the icon theme, enabling plugins…)
* It's annoying to have a popup without being able to check if the change was correct.

If 2 is chosen, it should be written that it requires a restart.

**Comment by goanpeca** [comment]
In that case of 2, I would make the popup appear at the end when clicking `ok`, or `apply`

**Comment by Nodd** [comment]
Yep. Bonus points if the dialog contains the reason why the restart is needed (as a list so that other settings can give their information too)

**Comment by goanpeca** [comment]
Yep, you read my mind ;-)

**Comment by goanpeca** [comment]
@Nodd ok, this is now updated to ask only after pressing Ok or Apply and displays a list of all the options that require a restart (and were changed).

Can you give it a try @ccordoba12? 

I would expect this to work in mac without changes but, @blink1073 can you give it a try?

**Comment by ccordoba12** [comment]
The restart popup message is very nice!! It looks very professional! Thanks a lot for working on it @goanpeca!

File: spyderlib/baseconfig.py [hunk 7]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
+    2.) Spyder provides ('en',  'fr', 'es' and 'pt_BR'), if the locale is
+    either 'pt' or 'pt_BR', this function will return 'pt_BR'
+    """
+    locale_language = locale.getdefaultlocale()[0]
+
+    if locale_language is None:
+        language = DEFAULT_LANGUAGE
+    else:
+        spyder_languages = get_available_translations()
+        for spyder_language in spyder_languages:
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Let's simplify the variables here to make it a bit easier to read. What about this?

```python
for lang in spyder_languages
```

**Comment by goanpeca** [comment]
Glad you like it, if this works on Mac, then it should be ready to merge

File: spyderlib/baseconfig.py [hunk 8]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
+            if locale_language == spyder_language:
+                language = locale_language
+                break
+            elif locale_language.startswith(spyder_language) or\
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Missing space: `or\` -> `or \`

File: spyderlib/baseconfig.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: Ok sure, I will adapt ;-)

File: spyderlib/baseconfig.py (hunk 8 above)
Comment (goanpeca) [inline spyderlib/baseconfig.py]: will fix, but probably will be one line with your suggestion

File: spyderlib/baseconfig.py [hunk 9]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
+                    spyder_language.startswith(locale_language):
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: 2 indentation spaces here, not 8, please (I know PEP8, but I don't like it)

File: spyderlib/baseconfig.py [hunk 10]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
...
-    # fixup environment var LANG in case it's unknown
-    if "LANG" not in os.environ:
-        import locale
-        lang = locale.getdefaultlocale()[0]
-        if lang is not None:
-            os.environ["LANG"] = lang
+
+    # fixup environment var LANG, LANGUAGE
+    language = load_lang_conf()
+    os.environ["LANG"] = language      # Works in windows
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: `works in windows` -> `Works on Windows`

The same for Ubuntu below, but changing it to Linux :-)

Comment (goanpeca) [inline spyderlib/baseconfig.py]: Is this going to work on Mac for sure?

Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Translations on Mac never have worked because Mac doesn't define proper `locale` values.

This work will let users select the translation they want to use :-)

File: spyderlib/plugins/configdialog.py [hunk 11]
```diff
@@ -90,6 +91,18 @@ def apply_changes(self):
...
+            # is needed before loading CONF, this is an extra method needed to
+            # ensure that when changes are applied, they are copied to a
+            # specific file storing the language value.
+            if getattr(self, 'save_lang', False):
+                self.save_lang()
+
+            for restart_option in self.restart_options:
+                if restart_option in self.changed_options:
+                    self.prompt_restart_required()
+                    break
```
Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: Why this `break` here?

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: I guess that I am under the impression that if I do not break it here, you will get a popup for every option that the user changed (and requires restart)

Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: I see, it's fine then :-)

File: spyderlib/plugins/configdialog.py [hunk 12]
```diff
@@ -661,13 +680,42 @@ def get_icon(self):
     def apply_settings(self, options):
         raise NotImplementedError
 
+    def prompt_restart_required(self):
+        """Prompt the user with a request to restart."""
+        restart_opt = self.restart_options
+        changed_opt = self.changed_options
```
Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: `restart_opt` -> `restart_opts` (because it's plural)
`changed_opt` -> `changed_opts`

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: fixed

File: spyderlib/plugins/configdialog.py [hunk 13]
```diff
@@ -704,6 +752,13 @@ def setup_page(self):
         margins_layout.addWidget(margin_spin)
         prompt_box = newcb(_("Prompt when exiting"), 'prompt_on_exit')
 
+        # Language chooser
+        langs = get_available_translations()
+        choices = list(zip([LANGUAGE_CODES[key] for key in langs], langs))
```
Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: What this `choices` variable hold? I mean, is it something like

```python
[('en': 'English', 'es': 'Spanish')]
```

?

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: I copied this from code above (the choices part and adapted it...) 

It holds what is stored in `LANGUAGE_CODES` in the `baseconfig.py` 

So the combo box display Español, English... etc... and the value is ...es, en... etc..

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: So actually it is more like 

`choices = [(English, 'en'), ('Español', 'es'), ...]`

File: spyderlib/baseconfig.py [hunk 14]
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
 #==============================================================================
 # Translations
 #==============================================================================
+LANG_FILE = osp.join(get_conf_path(), 'langconfig')
```
Comment (ccordoba12) [inline spyderlib/baseconfig.py]: Please rewrite this line as

```python
LANG_FILE = get_conf_path('langconfig')
```

That's the way `get_conf_path` works :-)

Comment (goanpeca) [inline spyderlib/baseconfig.py]: Done

File: spyderlib/plugins/configdialog.py (hunk 13 above)
Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: This seems a bit too complex to my taste (I mean, the difference between `LANGUAGE_CODES` and `get_available_translations`).

What will happen when someone adds `pt_PT` or `en_UK`?

**Comment by ccordoba12** [comment]
Everything seems fine to me now, but:

1. I still find the implementation too complex to follow.
2. We still need to test this PR on Mac :-)

File: spyderlib/plugins/configdialog.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: When someone adds this where?

**Comment by goanpeca** [comment]
Regarding 1.) hmmm

I tried to add all the needed documentation to understand. I will explain and you tell me where you find it confusing.

When someone wants to add a new translation they will create the folder in `locale` and put there the new translation, or variation of a translation.

Once they do this and run the code they will get an error because we need to have an updated language name (in native language) added to `LANGUAGE_CODES`. If they added a new language, no problem they just add i.e. `'de': 'Deutsch'`... if it is a variation they just add i.e. `'pt_PT': 'Portuguese (Portugal)'` and make sure the other variation (if existent) should be then `'pt_BR': 'Portuguese (Brazil)'`

But the truth is, that the second case will most likely never happen...

File: spyderlib/plugins/configdialog.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: I could just use the `LANGUAGE_CODES` variable and remove the call to `langs = get_available_translations()`

File: spyderlib/plugins/configdialog.py [hunk 15]
```diff
@@ -704,6 +752,12 @@ def setup_page(self):
         margins_layout.addWidget(margin_spin)
         prompt_box = newcb(_("Prompt when exiting"), 'prompt_on_exit')
 
+        # Language chooser
+        choices = sorted([(val, key) for key, val in LANGUAGE_CODES.items()])
+        language_combo = self.create_combobox(_('Language'), choices,
+                                              'interface_language',
+                                              restart=True)
+
```
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: @ccordoba12 this should be clearer now

Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: Yep, much better, thanks! Now I don't have objecttions.

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: :+1:

**Comment by ccordoba12** [comment]
I'll merge this tomorrow, after testing on Mac :-)

**Comment by goanpeca** [comment]
I will have to buy a Mac... ;-)

**Comment by blink1073** [comment]
I can test today on my Mac if you prefer @ccordoba12.

**Comment by ccordoba12** [comment]
That'd be great too @blink1073! I thought you weren't around ;-)

**Comment by blink1073** [comment]
Was at the beach with the little ones.  ;-)

**Comment by blink1073** [comment]
Una problema:

```python
Cargando el inspector de objetos...
Traceback (most recent call last):
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 3077, in main
    mainwindow = run_spyder(app, options, args)
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 2963, in run_spyder
    main.setup()
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 851, in setup
    self.set_splash(_("Loading outline explorer..."))
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 2095, in set_splash
    self.debug_print(message)
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 532, in debug_print
    debug_print(message)
  File "/Users/silvester/workspace/spyder/spyderlib/baseconfig.py", line 56, in debug_print
    print(*message, file=ss)
UnicodeEncodeError: 'ascii' codec can't encode character '\xf3' in position 27: ordinal not in range(128)
```

**Comment by blink1073** [comment]
It looks like `sys.stdout` is encoded as `ascii`, thanks Apple...

**Comment by goanpeca** [comment]
Any suggestions?

**Comment by blink1073** [comment]
Looking into it now.

**Comment by blink1073** [comment]
Hmm, this works fine from an IPython prompt:  `print('€', file=sys.stdout)`

**Comment by blink1073** [comment]
It might have something to do with the restart itself affecting `sys.stdout`.

**Comment by goanpeca** [comment]
If you say No when the popup box asks to restart... and then you close spyder...
and rerun from bootstrap...?

You get the same error?

**Comment by blink1073** [comment]
If I restart after the failure itself it comes up fine.

**Comment by goanpeca** [comment]
If you run without DEBUG ?

**Comment by blink1073** [comment]
Actually if I say NO then it restarts anyway...

**Comment by goanpeca** [comment]
hmm thats a bug then.. it should do that!

**Comment by blink1073** [comment]
Oops, forget that last comment.

**Comment by goanpeca** [comment]
Can you run without Debug? and check again?

**Comment by blink1073** [comment]
Running without debug is fine, and I can print unicode characters from the Spyder Python console.

**Comment by blink1073** [comment]
It appears to just be the restart while in debug mode.

**Comment by blink1073** [comment]
I tried using `sys.stdout` within `debug_print`, instead of the global variable, but it still did not work.

**Comment by goanpeca** [comment]
Can you try this...

Change baseconfig.py  line 52 with this

```python
def debug_print(*message):
    """Output debug messages to stdout"""
    if DEBUG:
        ss = STDOUT
        message = tuple([m.encode('utf-8') for m in message])
        print(*message, file=ss)
```

And try with debug again?

**Comment by blink1073** [comment]
That works but all of the strings print with `b''`

**Comment by blink1073** [comment]
E.g.: `b'Configuration de la fen\xc3\xaatre principale...'`

**Comment by goanpeca** [comment]
and if you do this

```python
def debug_print(*message):
    """Output debug messages to stdout"""
    if DEBUG:
        ss = STDOUT
        message = tuple([to_text_string(m) for m in message])
        print(*message, file=ss)
```

**Comment by blink1073** [comment]
Nope, same UnicodeEncodeError.

**Comment by goanpeca** [comment]
You are on Py3?

**Comment by blink1073** [comment]
Yep, I'll try Py2.

**Comment by blink1073** [comment]
Same problem there.

**Comment by blink1073** [comment]
If I turn DEBUG off, do the restart, and then try and print from with the Spyder Internal Console, I get the same error too.  The other consoles are note affected.

**Comment by goanpeca** [comment]
yes... the internal console seems to be the problem...

But that is something unrelated to the changes I did, seems to be a problem that has been sleeping for some time. I am trying some options

**Comment by blink1073** [comment]
Yes, it appears to be related to the restart itself.

**Comment by blink1073** [comment]
Yup, confirmed.  If I start in English mode and do just a restart, we get this behavior.

**Comment by goanpeca** [comment]
I will try linux.. one sec

**Comment by goanpeca** [comment]
Ok, I can reproduce it, ...fixing!! give me 10 mins

**Comment by goanpeca** [comment]
@blink1073 

Can you add this line to `spyderlib\spyder.py` after line 2737

`env['PYTHONIOENCODING'] = 'utf-8'`

**Comment by blink1073** [comment]
Perfecto!

**Comment by goanpeca** [comment]
both py2 and py3?

**Comment by blink1073** [comment]
Oui.

**Comment by blink1073** [comment]
`debug_print` is super nice, I use it all the time.

**Comment by Nodd** [comment]
This env variable will change the behavior in scripts run from spyder ? This is bad, we had many bug reports because of this. We should change the encoding of the internal console only. I think I have a solution somewhere, I'll try to find it.

**Comment by goanpeca** [comment]
@Nodd I do not think it will be the case... but you can try with an example?

**Comment by goanpeca** [comment]
I can erase the environment variable in the `restart_app.py` just to be sure... so it is only used for the python process that spawns the new Spyder instance.

**Comment by goanpeca** [comment]
Maybe I can pop that variable in `spyderlib\spyder.py` so that it is not there after startup. It is needed when spawning the new Spyder process on restart, but should not be needed afterwards.

**Comment by blink1073** [comment]
But what if the user had explicitly set that variable to something else?  Perhaps only set it if it is not in `os.environ` and not set to something else, then leave it there.

**Comment by blink1073** [comment]
`utf-8` all the things!

**Comment by goanpeca** [comment]
Well, if the user set it to soemthing else, the spyder will break...

**Comment by blink1073** [comment]
Then I say set it and leave it.

**Comment by goanpeca** [comment]
Lets see what @Nodd brings :-)

**Comment by SylvainCorlay** [comment]
I am good with this. Will use that dialog on restarting spyder for the icons as soon as this is merged.

**Comment by goanpeca** [comment]
@Nodd did you have some time to find what you were looking for? [or](https://www.youtube.com/watch?v=EabB55IYyOg)

@ccordoba12  any comments on the fix needed to get this PR working properly, specifically [this last commit](https://github.com/goanpeca/spyder/commit/5bae10be0773ab2a3b3a2a943606ab49e4d69d96)

**Comment by Nodd** [comment]
I used this in some scritps so that `argparse` help doesn't raise a `UnicodeSomethingError` when displaying a help message with utf8 symbols under python2:
```
import codecs
sys.stdout = codecs.getwriter('utf-8')(sys.stdout)
```
Since the internal console is basically `sys.stdout`, it might be used as-is.

**Comment by goanpeca** [comment]
and in python 3 ?

**Comment by goanpeca** [comment]
where should I put this code?

**Comment by Nodd** [comment]
In python3 I didn't have this problem (the unicode support is way better).
It should be put at the same place where sys.stdout is redirected to the internal console, I guess. I have no idea where it is :disappointed:

File: spyderlib/spyder.py [hunk 16]
```diff
@@ -2743,6 +2743,9 @@ def restart(self):
             startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
             shell = False
         else:
+            # This environment variable is needed so that the new python
+            # process spawned with Popen sets this as the default encoding
+            env['PYTHONIOENCODING'] = 'utf-8'
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: I think setting this system-wide is not a good idea. For example, It could mess with Pandas trying to import/export files :-)

**Comment by ccordoba12** [comment]
If I'm not wrong, redirection of stdio for the internal console happens in `redirect_internalshell_stdio`, currently present in

https://github.com/spyder-ide/spyder/blob/master/spyderlib/spyder.py#L2435

**Comment by goanpeca** [comment]
I tried putting @Nodd  where @ccordoba12 without success

**Comment by goanpeca** [comment]
@Nodd I think I have it fixed, can you check the latest commit?

@blink1073, or @ccordoba12  can you please test on Mac? remember to run using 

`python bootstrap.py --debug`

**Comment by blink1073** [comment]
It works for me, other than the fact that the restart dialog was entirely in English.

**Comment by blink1073** [comment]
screen shot 2015-05-07 at 8 26 03 pm

**Comment by goanpeca** [comment]
Cause we need To rerun the localización scripts :+1:

**Comment by goanpeca** [comment]
@ccordoba12 can we merge this one so that @SylvainCorlay can add this to his icon PR :smile: ?

**Comment by SylvainCorlay** [comment]
:+1:

**Comment by SylvainCorlay** [comment]
Yeah seems I will need to rebase again anyway!

File: spyderlib/config.py [hunk 17]
```diff
@@ -169,6 +169,7 @@ def is_ubuntu():
               'animated_docks': True,
               'prompt_on_exit': False,
               'panes_locked': True,
+              'interface_language': load_lang_conf(),
```
Comment (ccordoba12) [inline spyderlib/config.py]: I think we don't need to add this here because this is dependent on the user system.

The only options that we add in `config.py` are those that can be changed and/or updated in the future (by bumping `CONF_VERSION`). That's why (for example) we don't add here the window size or state.

Comment (ccordoba12) [inline spyderlib/config.py]: Another example of things we don't add are the run configuration options per file :-)

File: spyderlib/plugins/configdialog.py [hunk 18]
```diff
@@ -90,8 +91,20 @@ def apply_changes(self):
             self.save_to_conf()
             if self.apply_callback is not None:
                 self.apply_callback()
+
+            # Since the language cannot be retrieved by CONF and the language
+            # is needed before loading CONF, this is an extra method needed to
+            # ensure that when changes are applied, they are copied to a
+            # specific file storing the language value.
+            if getattr(self, 'save_lang', False):
```
Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: Why is this line necessary?

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: Cause this class is inherited by all config panes. But only the main will for sure have the save_lang method.

File: spyderlib/config.py (hunk 17 above)
Comment (ccordoba12) [inline spyderlib/config.py]: Please remove this option from here to finally merge this PR :-)

If I understand the code correctly:

1. It'll be saved to `spyder.ini` by the `save_lang` method of `MainConfigPage`
2. It can't be retrieved by `CONF` directly.

So there's no need to have it here.

Comment (goanpeca) [inline spyderlib/config.py]: @ccordoba12 

If I don't add that method to the config, Spyder will not be able to load the preferences after a reset, cause there will be no `interface_language` setting in the .ini file.

Suggestions?

Comment (goanpeca) [inline spyderlib/config.py]: I think we should leave it as it is, otherwise we would have to made unneded changes either to how the preferences are loaded or do an extra check in spyder.y to see if this is the first time it runs.... here it is just one line of code...

File: spyderlib/plugins/configdialog.py (hunk 18 above)
Comment (Nodd) [inline spyderlib/plugins/configdialog.py]: how about changing the name to `__save_lang` ? this way it will be "hidden" to the child class (external plugins included), but the name will be changed at runtime (the class name is added).

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: sure

**Comment by SylvainCorlay** [comment]
Is there any blocking point for this one?

File: spyderlib/config.py [hunk 19]
```diff
@@ -724,6 +724,9 @@ def is_ubuntu():
 CONF = UserConfig('spyder', defaults=DEFAULTS, load=True, version=CONF_VERSION,
                   subfolder=SUBFOLDER, backup=True, raw_mode=True)
 
+# Ensures that the config is present on spyder first run
+CONF.set('main', 'interface_language', load_lang_conf())
```
Comment (goanpeca) [inline spyderlib/config.py]: @ccordoba12 based on your previous comment I need to add this line here to make sure that on first run of spyder the `interface_language` entry is created. On subsequent starts of spyder that line is harmless so this way is not part of the CONF but ensures the correct behavior. 

You think is ok like this?

**Comment by goanpeca** [comment]
@ccordoba12 this one should be ready to merge as well

**Comment by SylvainCorlay** [comment]
:+1:

File: spyderlib/config.py (hunk 19 above)
Comment (ccordoba12) [inline spyderlib/config.py]: Yes, I think it's fine :-) But I think it's better to move it `configdialog.py`. Well, if it's possible because there is where this setting is really needed.

File: spyderlib/plugins/configdialog.py [hunk 20]
```diff
@@ -90,8 +91,21 @@ def apply_changes(self):
             self.save_to_conf()
             if self.apply_callback is not None:
                 self.apply_callback()
+
+            # Since the language cannot be retrieved by CONF and the language
+            # is needed before loading CONF, this is an extra method needed to
+            # ensure that when changes are applied, they are copied to a
+            # specific file storing the language value. This only applies to
+            # the main section config.
+            if self.CONF_SECTION == u'main':
```
Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: I don't understand this change here. Maybe it's better to just move the `CONF.set` line from `config.py` here to avoid this.

Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: I mean, what I don't understand is why you're looking for `main`.

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: `main` is the section in charge of setting the interface language, so the `_save_lang` only belongs to that section, before I was using  `if getattr(self, "_save_lang", False):` but I change it so that it was more explicit.

Since this method runs for every child (any other section or plugin) then I need this to avoid an error of calling _save_lang when it does not exist.

File: spyderlib/config.py (hunk 19 above)
Comment (goanpeca) [inline spyderlib/config.py]: Hmmm ok I will take a look if I find a suitable place there

File: spyderlib/plugins/configdialog.py (hunk 20 above)
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: The other way would be to make the save_lang method part of one of the ancestors so that EVERY section would alwyas make that save... but since it is only necessary for main...

Comment (ccordoba12) [inline spyderlib/plugins/configdialog.py]: Too much work :-) Then leave this part as it is now.

Please see if you can move the `CONF.set` call from `config.py` to here so that we can (finally!) merge this PR ;-)

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: Will do! ;-)

File: spyderlib/plugins/configdialog.py [hunk 21]
```diff
@@ -151,6 +165,9 @@ def __init__(self, parent=None):
 
         self.setWindowTitle(_("Preferences"))
         self.setWindowIcon(get_icon("configure.png"))
+
+        # Ensures that the config is present on spyder first run
+        CONF.set('main', 'interface_language', load_lang_conf())
```
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: Moved here, right at the end of the init :-), should work fine now

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: @ccordoba12

**Comment by goanpeca** [comment]
Merging :-) @ccordoba12?

**Comment by ccordoba12** [comment]
Yes!! I'm having a problem with Qt5 but it seems to be with my system rather than with this PR (because I'm also having problems in master :-)

**Comment by ccordoba12** [comment]
Thanks for your patience by the way ;-)

**Comment by goanpeca** [comment]
No prob ;-)

Ok, @SylvainCorlay, now is your turn :-)  

Just add `restart=True` to the combobox, i.e.

```python
        language_combo = self.create_combobox(_('Language'), choices,
                                              'interface_language',
                                              restart=True)
```
